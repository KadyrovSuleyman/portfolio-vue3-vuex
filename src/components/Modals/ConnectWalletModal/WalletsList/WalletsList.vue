<script setup lang="ts">

import { computed } from 'vue';
import Div from '@/elements/Div/Div.vue';
import propsObj from '@/elements/propsObj';
import { useStore } from 'vuex';
import adaptState from './state';
import WalletsListItem from './WalletsListItem/WalletsListItem.vue';
import { clickHandlerGenerator } from './logic';

// eslint-disable-next-line no-undef
const props = defineProps({ ...propsObj });
const comp = computed(() => ({ elem: props.elem || 'walletsList' }));

const store = useStore();
const state = computed(() => adaptState(store));

const clickHandler = clickHandlerGenerator(
  store,
  `${props.block ? `${props.block}-` : ''}${comp.value.elem}`,
);

</script>

<template>
  <Div :block="props.block" :elem="comp.elem" :mods="props.mods"
    :onClick="clickHandler"
  >

    <WalletsListItem v-for="{ name, icon } in state.walletsList" :key="name"
      :icon="icon" :block="comp.elem"
    >
      {{ name }}
    </WalletsListItem>

  </Div>
</template>

<style lang="scss">
  @use 'styles' as *;
</style>
