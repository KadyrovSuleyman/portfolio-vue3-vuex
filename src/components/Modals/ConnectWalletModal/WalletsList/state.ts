import { Store } from 'vuex';

const adaptState = (store: Store<any>) => ({
  walletsList: store.state.wallet.list,
});
export default adaptState;
