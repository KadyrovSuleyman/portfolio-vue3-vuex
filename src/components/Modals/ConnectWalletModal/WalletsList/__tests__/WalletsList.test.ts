import { mount, VueWrapper } from '@vue/test-utils';
import { createStore, Store } from 'vuex';
import WalletsList from '../WalletsList.vue';

jest.mock('../state.ts');

let store: Store<any>;
let wrapper: VueWrapper<any>;

beforeEach(() => {
  store = createStore({
    state: {
      walletsList: [
        {
          name: 'MetaMask',
          icon: 'icon/metaMask.svg',
        },
        {
          name: 'Walletconnect',
          icon: 'icon/walletconnect.svg',
        },
      ],
    },
    mutations: {
      add: (state, item: any) => { state.walletsList.push(item); },
      delete: (state, index: number) => { state.walletsList.splice(index, 1); },
      change: (state, index: number) => {
        state.walletsList[index].name = 'changed name';
        state.walletsList[index].icon = 'changed icon';
      },
    },
  });
});
afterEach(() => {
  store = createStore({});
  wrapper.unmount();
});

// ===========================

it('walletsList renders', () => {
  wrapper = mount(WalletsList, { global: { plugins: [store] } });
  expect(wrapper.element.outerHTML).toMatchSnapshot();
});

it('watchs props changes', async () => {
  wrapper = mount(WalletsList, { global: { plugins: [store] } });
  expect(wrapper.find('.walletsList').classes()).toEqual(['walletsList']);

  await wrapper.setProps({
    mods: {
      selected: true,
    },
  });
  expect(wrapper.find('.walletsList').classes()).toEqual([
    'walletsList',
    'walletsList__selected',
  ]);

  await wrapper.setProps({
    mods: {
      selected: false,
    },
  });
  expect(wrapper.find('.walletsList').classes()).toEqual(['walletsList']);
});

describe('outer store changing', () => {
  it('adding item', async () => {
    wrapper = mount(WalletsList, {
      global: { plugins: [store] },
    });

    expect(wrapper.findAll('.walletsList-walletsListItem').length).toBe(2);

    store.commit('add', {
      name: 'scum name',
      icon: 'scum icon',
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.findAll('.walletsList-walletsListItem').length).toBe(3);
    expect(wrapper.element.outerHTML).toMatchSnapshot();
  });

  it('deleting item', async () => {
    wrapper = mount(WalletsList, {
      global: { plugins: [store] },
    });

    expect(wrapper.findAll('.walletsList-walletsListItem').length).toBe(2);

    store.commit('delete', 1);
    await wrapper.vm.$nextTick();
    expect(wrapper.findAll('.walletsList-walletsListItem').length).toBe(1);

    store.commit('delete', 0);
    await wrapper.vm.$nextTick();
    expect(wrapper.findAll('.walletsList-walletsListItem').length).toBe(0);
  });

  it('changing item', async () => {
    wrapper = mount(WalletsList, {
      global: { plugins: [store] },
    });

    expect(wrapper.findAll('.walletsList-walletsListItem').length).toBe(2);

    store.commit('change', 0);
    await wrapper.vm.$nextTick();
    expect(wrapper.findAll('.walletsList-walletsListItem').length).toBe(2);
    expect(wrapper.element.outerHTML).toMatchSnapshot();
  });
});
