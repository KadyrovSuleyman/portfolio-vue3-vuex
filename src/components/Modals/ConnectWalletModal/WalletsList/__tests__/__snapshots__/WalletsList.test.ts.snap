// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`outer store changing adding item 1`] = `
<div class="walletsList">
  <div class="walletsList-walletsListItem"><img class="walletsListItem-icon" alt="" src="icon/metaMask.svg"><span class="walletsListItem-span">MetaMask</span></div>
  <div class="walletsList-walletsListItem"><img class="walletsListItem-icon" alt="" src="icon/walletconnect.svg"><span class="walletsListItem-span">Walletconnect</span></div>
  <div class="walletsList-walletsListItem"><img class="walletsListItem-icon" alt="" src="scum icon"><span class="walletsListItem-span">scum name</span></div>
</div>
`;

exports[`outer store changing changing item 1`] = `
<div class="walletsList">
  <div class="walletsList-walletsListItem"><img class="walletsListItem-icon" alt="" src="changed icon"><span class="walletsListItem-span">changed name</span></div>
  <div class="walletsList-walletsListItem"><img class="walletsListItem-icon" alt="" src="icon/walletconnect.svg"><span class="walletsListItem-span">Walletconnect</span></div>
</div>
`;

exports[`walletsList renders 1`] = `
<div class="walletsList">
  <div class="walletsList-walletsListItem"><img class="walletsListItem-icon" alt="" src="icon/metaMask.svg"><span class="walletsListItem-span">MetaMask</span></div>
  <div class="walletsList-walletsListItem"><img class="walletsListItem-icon" alt="" src="icon/walletconnect.svg"><span class="walletsListItem-span">Walletconnect</span></div>
</div>
`;
